File: Graphs/open_lock.py
from collections import deque
from typing import List


class OpenLock:
    def openLock(self, deadends: List[str], target: str) -> int:
        seen = set(deadends)
        if "0000" in seen:
            return -1
        seen.add("0000")
        queue = deque(["0000"])
        steps = 0
        while queue:
            for _ in range(len(queue)):
                combo = queue.popleft()
                if combo == target:
                    return steps

                combo_list = list(combo)
                for i in range(4):
                    digit = int(combo_list[i])
                    for change in [-1, 1]:
                        new_digit = digit + change
                        if new_digit == -1:
                            new_digit = 9
                        if new_digit == 10:
                            new_digit = 0
                        new_combo = combo_list.copy()
                        new_combo[i] = str(new_digit)
                        new_combo = "".join(new_combo)
                        if new_combo not in seen:
                            seen.add(new_combo)
                            queue.append(new_combo)

            steps += 1

        return -1


# Time complexity: O(10^4) since there are 10,000 possible combinations of the lock.
# Space complexity: O(10^4) since we may have to store all combinations in the seen set and the queue.

File: Graphs/openLock.js
const openLock = (deadends, target) => {
  const seen = new Set(deadends);

  if (seen.has("0000")) {
    return -1;
  }

  seen.add("0000");

  let queue = ["0000"];
  let turns = 0;

  while (queue.length) {
    const nextQueue = [];
    for (let combination of queue) {
      if (combination === target) return turns;

      for (let i = 0; i < 4; i++) {
        const digit = parseInt(combination[i], 10);
        const moves = [1, -1];
        for (let change of moves) {
          // JavaScript modulo can yield negative results, so we add 10 to ensure it stays positive
          const newDigit = (digit + change + 10) % 10;
          const newCombination =
            combination.slice(0, i) + newDigit + combination.slice(i + 1);
          if (!seen.has(newCombination)) {
            seen.add(newCombination);
            nextQueue.push(newCombination);
          }
        }
      }
    }

    queue = nextQueue;
    turns++;
  }

  return -1;
};

// Time Complexity: O(10^4) - In the worst case, we may need to explore all
// possible combinations of the lock (from "0000" to "9999").
// Space Complexity: O(10^4) - We may need to store all possible combinations in
// the seen set and the queue in the worst case.
